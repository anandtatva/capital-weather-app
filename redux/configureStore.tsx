import { createStore, combineReducers } from "redux";
import { persistReducer, persistStore } from "redux-persist"
import AsyncStorage from "@react-native-async-storage/async-storage"
import CountrySlice from "./reducer/country-slice"
const persistConfig = {
    key: "country",
    storage: AsyncStorage
}

const rootReducer = combineReducers({
    country: CountrySlice
})

const persistReducers = persistReducer(persistConfig, rootReducer)

const store = createStore(persistReducers);

const persistor = persistStore(store)

export {
    store,
    persistor
}