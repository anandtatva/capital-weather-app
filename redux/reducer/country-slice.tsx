import { createSlice } from "@reduxjs/toolkit";
import { CountryType } from "../../types/country-type"
import { CapitalWeatherType } from "../../types/capital-weather-type"


interface CountryState {
    name: string;
    countries: CountryType[];
    capitalWeather: CapitalWeatherType | null;
}

const initialState: CountryState = {
    name: "",
    capitalWeather: null,
    countries: []
}

const CountrySlice = createSlice({
    name: "country-slice",
    initialState,
    reducers: {
        setName: (state, action) => {
            state.name = action.payload
        },
        setCountries: (state, action) => {
            state.countries = action.payload
        },
        setCapitalWeather: (state, action) => {
            state.capitalWeather = action.payload
        },

    }
})


export const { setCapitalWeather, setCountries, setName } = CountrySlice.actions;
export default CountrySlice.reducer;