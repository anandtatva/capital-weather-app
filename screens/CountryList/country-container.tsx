import { View, Image, Text, StyleSheet, FlatList } from "react-native";
import { useSelector, useDispatch } from "react-redux"
import { CountryType } from "../../types/country-type";
import Button from "../../components/Button";
import { colors } from "../../utils";

interface CountryContainerProps {
    index: number;
    item: CountryType;
    onPress: (item: CountryType) => void;
}
const CountryContainer = ({ index, item, onPress }: CountryContainerProps) => {
    const handleWeather = async () => {
        await onPress(item)
    }

    return (
        <View style={styles.container}>
            <View style={{ flex: 1, padding: 5 }}>
                <Image style={styles.image} source={{ uri: item.flag }} />
            </View>
            <View style={styles.textContainer}>
                <Text style={styles.text} testID={`Capital-${index}`}>Capital: {item.capital}</Text>
                <Text style={styles.text} testID={`Population-${index}`}>Population: {item.population} </Text>
                <Text style={styles.text} testID={`LatLng-${index}`}>LatLng: {item.latlng.join(",")}</Text>
                <Button
                    rootStyle={styles.button}
                    btnTextStyle={styles.btnText}
                    onButtonPress={handleWeather}
                    name={"Captal Weather"}
                    testId={`Captal-Weather-Button-${index}`}
                />
            </View>
        </View>
    )
}

export default CountryContainer;

const styles = StyleSheet.create({
    button: {
        backgroundColor: colors.secondary,
        marginVertical: 5
    },
    btnText: {
        color: colors.primary
    },
    container: {
        flexDirection: "row",
        padding: 10,
        backgroundColor: colors.primary
    },
    image: {
        width: 25,
        height: 25,
        borderRadius: 50
    },
    text: {
        color: colors.secondary,
        lineHeight: 20
    },
    textContainer: {
        flex: 4

    }
})