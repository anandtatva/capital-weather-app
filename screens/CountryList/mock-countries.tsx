import { CountryType } from "../../types/country-type";

export const mockContries: CountryType[] = [
    {
        capital: 'Diego Garcia',
        latlng: [-6, 71.5],
        flag: 'https://flagcdn.com/io.svg',
        population: 3000
    },
    {
        capital: 'New Delhi',
        latlng: [20, 77],
        flag: 'https://flagcdn.com/in.svg',
        population: 1380004385
    }
]