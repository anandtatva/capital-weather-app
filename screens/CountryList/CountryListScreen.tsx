import { View, Text, StyleSheet, FlatList } from "react-native";
import { useSelector, useDispatch } from "react-redux"
import { CountryType } from "../../types/country-type";
import CountryContainer from "./country-container";
import { colors, showAlert } from "../../utils";
import { capitalWeatherAPIs } from "../../services/api";
import { setCapitalWeather } from "../../redux/reducer/country-slice";

const CountryListScreen = ({ navigation }) => {
    const country = useSelector(state => state.country);
    const dispatch = useDispatch();

    const handleCapitalWeather = async (item: CountryType) => {
        const resp = await capitalWeatherAPIs(item.capital);
        if (resp.status) {
            dispatch(setCapitalWeather(resp.data))
            navigation.navigate("CapitalWeatherScreen")
        } else showAlert(resp.message)
    }
    const renderItem = ({ index, item }: { item: CountryType; index: number; }) => (
        <CountryContainer index={index} item={item} onPress={handleCapitalWeather} />
    )
    return (
        <View>
            <FlatList
                data={country?.countries || []}
                renderItem={renderItem}
                keyExtractor={(_, index) => String(index)}
                ItemSeparatorComponent={() => <View style={styles.separator} />}
            />
        </View>
    )
}

export default CountryListScreen;

const styles = StyleSheet.create({
    container: {

    },
    separator: {
        borderBottomWidth: 1,
        borderBottomColor: colors.secondary
    }
})