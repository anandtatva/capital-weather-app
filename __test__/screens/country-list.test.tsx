import { render, waitFor, cleanup, fireEvent, act } from "@testing-library/react-native";
import { Provider } from "react-redux";
import { store } from "../../redux/configureStore";
import CountryListScreen from "../../screens/CountryList/CountryListScreen";
import { setName, setCountries, setCapitalWeather } from "../../redux/reducer/country-slice";
import { Alert } from "react-native";
import { mockContries } from "../../screens/CountryList/mock-countries";

jest.mock("@react-navigation/native", () => {
    const actualNav = jest.requireActual("@react-navigation/native")
    return {
        ...actualNav,
        useFocusEffect: () => jest.fn(),
        useNavigation: () => ({
            navigate: jest.fn()
        }),
        useRoute: () => ({
            params: {}
        })
    }
})

const spyAlert = jest.spyOn(Alert, "alert")

describe("Screen : Country List", () => {
    beforeEach(() => {
        fetchMock.mockIf(/^https?:\/\/example.com.*S/)
        store.dispatch(setName(""));
        store.dispatch(setCountries([]))
        store.dispatch(setCapitalWeather(null))
    })
    afterEach(() => {
        jest.clearAllMocks();
        cleanup()
    })
    it("renders correctly", () => {
        const navigation = { navigate: jest.fn() }
        const tree = render(
            <Provider store={store}>
                <CountryListScreen navigation={navigation} />
            </Provider>
        ).toJSON()
        expect(tree).toMatchSnapshot();
    })
    it("Countries should be render correctly", () => {
        const navigation = { navigate: jest.fn() }
        store.dispatch(setCountries(mockContries))
        const tree = render(
            <Provider store={store}>
                <CountryListScreen navigation={navigation} />
            </Provider>
        ).toJSON()
        expect(tree).toMatchSnapshot();
    })

    it("Valid country details should be  render in country container", () => {
        const navigation = { navigate: jest.fn() }
        store.dispatch(setCountries(mockContries))
        const { getByTestId } = render(
            <Provider store={store}>
                <CountryListScreen navigation={navigation} />
            </Provider>
        );
        const capital = getByTestId("Capital-0");
        const population = getByTestId("Population-0");
        const latLng = getByTestId("LatLng-0");
        expect(capital.children[1]).toBe(mockContries[0].capital)
        expect(population.children[1]).toBe(String(mockContries[0].population))
        expect(latLng.children[1]).toBe(mockContries[0].latlng.join(","))
    })
    it("Capital weather should be show on press capital weather button", async () => {
        const navigation = { navigate: jest.fn() }
        store.dispatch(setCountries(mockContries))
        const { getByTestId } = render(
            <Provider store={store}>
                <CountryListScreen navigation={navigation} />
            </Provider>
        );
        await act(async () => {
            const capitalWeatherButton0 = getByTestId("Captal-Weather-Button-0");
            await fireEvent.press(capitalWeatherButton0)
            expect(navigation.navigate).toHaveBeenCalledWith("CapitalWeatherScreen")
            expect(Boolean(store.getState().country.capitalWeather)).toBeTruthy();
        })

    })
})