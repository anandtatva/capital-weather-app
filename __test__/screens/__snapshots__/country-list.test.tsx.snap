// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Screen : Country List Countries should be render correctly 1`] = `
<View>
  <RCTScrollView
    ItemSeparatorComponent={[Function]}
    data={
      Array [
        Object {
          "capital": "Diego Garcia",
          "flag": "https://flagcdn.com/io.svg",
          "latlng": Array [
            -6,
            71.5,
          ],
          "population": 3000,
        },
        Object {
          "capital": "New Delhi",
          "flag": "https://flagcdn.com/in.svg",
          "latlng": Array [
            20,
            77,
          ],
          "population": 1380004385,
        },
      ]
    }
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    viewabilityConfigCallbackPairs={Array []}
  >
    <View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            Object {
              "backgroundColor": "#000",
              "flexDirection": "row",
              "padding": 10,
            }
          }
        >
          <View
            style={
              Object {
                "flex": 1,
                "padding": 5,
              }
            }
          >
            <Image
              source={
                Object {
                  "uri": "https://flagcdn.com/io.svg",
                }
              }
              style={
                Object {
                  "borderRadius": 50,
                  "height": 25,
                  "width": 25,
                }
              }
            />
          </View>
          <View
            style={
              Object {
                "flex": 4,
              }
            }
          >
            <Text
              style={
                Object {
                  "color": "#FFF",
                  "lineHeight": 20,
                }
              }
              testID="Capital-0"
            >
              Capital: 
              Diego Garcia
            </Text>
            <Text
              style={
                Object {
                  "color": "#FFF",
                  "lineHeight": 20,
                }
              }
              testID="Population-0"
            >
              Population: 
              3000
               
            </Text>
            <Text
              style={
                Object {
                  "color": "#FFF",
                  "lineHeight": 20,
                }
              }
              testID="LatLng-0"
            >
              LatLng: 
              -6,71.5
            </Text>
            <View
              accessible={true}
              collapsable={false}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "backgroundColor": "#FFF",
                  "marginVertical": 5,
                  "opacity": 1,
                  "paddingVertical": 10,
                }
              }
              testID="Captal-Weather-Button-0"
            >
              <Text
                style={
                  Array [
                    Object {
                      "color": "#FFF",
                      "textAlign": "center",
                    },
                    Object {
                      "color": "#000",
                    },
                  ]
                }
              >
                Captal Weather
              </Text>
            </View>
          </View>
        </View>
        <View
          style={
            Object {
              "borderBottomColor": "#FFF",
              "borderBottomWidth": 1,
            }
          }
        />
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            Object {
              "backgroundColor": "#000",
              "flexDirection": "row",
              "padding": 10,
            }
          }
        >
          <View
            style={
              Object {
                "flex": 1,
                "padding": 5,
              }
            }
          >
            <Image
              source={
                Object {
                  "uri": "https://flagcdn.com/in.svg",
                }
              }
              style={
                Object {
                  "borderRadius": 50,
                  "height": 25,
                  "width": 25,
                }
              }
            />
          </View>
          <View
            style={
              Object {
                "flex": 4,
              }
            }
          >
            <Text
              style={
                Object {
                  "color": "#FFF",
                  "lineHeight": 20,
                }
              }
              testID="Capital-1"
            >
              Capital: 
              New Delhi
            </Text>
            <Text
              style={
                Object {
                  "color": "#FFF",
                  "lineHeight": 20,
                }
              }
              testID="Population-1"
            >
              Population: 
              1380004385
               
            </Text>
            <Text
              style={
                Object {
                  "color": "#FFF",
                  "lineHeight": 20,
                }
              }
              testID="LatLng-1"
            >
              LatLng: 
              20,77
            </Text>
            <View
              accessible={true}
              collapsable={false}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "backgroundColor": "#FFF",
                  "marginVertical": 5,
                  "opacity": 1,
                  "paddingVertical": 10,
                }
              }
              testID="Captal-Weather-Button-1"
            >
              <Text
                style={
                  Array [
                    Object {
                      "color": "#FFF",
                      "textAlign": "center",
                    },
                    Object {
                      "color": "#000",
                    },
                  ]
                }
              >
                Captal Weather
              </Text>
            </View>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
</View>
`;

exports[`Screen : Country List renders correctly 1`] = `
<View>
  <RCTScrollView
    ItemSeparatorComponent={[Function]}
    data={Array []}
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    viewabilityConfigCallbackPairs={Array []}
  >
    <View />
  </RCTScrollView>
</View>
`;
